plugins/gdk-pixbuf/pixbufloader-jxl.c

```diff
@@ -3,258 +3,257 @@
-    fprintf(stderr, "JxlDecoderSubscribeEvents failed\n");
-    JxlDecoderDestroy(dec);
-    return 0;
-  }
+  gboolean done;
+  gboolean alpha_premultiplied;
 
-  JxlBasicInfo info;
-  int success = 0;
-  JxlPixelFormat format = {4, JXL_TYPE_UINT8, JXL_NATIVE_ENDIAN, 0};
-  JxlDecoderSetInput(dec, next_in, size);
+  gpointer icc_buff;
+  skcms_ICCProfile icc;
 
-  for (;;) {
-    JxlDecoderStatus status = JxlDecoderProcessInput(dec);
+} GdkPixbufJXLDecoderState;
 
-    if (status == JXL_DEC_ERROR) {
-      fprintf(stderr, "Decoder error\n");
-      break;
-    } else if (status == JXL_DEC_NEED_MORE_INPUT) {
-      fprintf(stderr, "Error, already provided all input\n");
-      break;
-    } else if (status == JXL_DEC_BASIC_INFO) {
-      if (JXL_DEC_SUCCESS != JxlDecoderGetBasicInfo(dec, &info)) {
-        fprintf(stderr, "JxlDecoderGetBasicInfo failed\n");
-        break;
-      }
-      *xsize = info.xsize;
-      *ysize = info.ysize;
-      *stride = info.xsize * 4;
-    } else if (status == JXL_DEC_NEED_IMAGE_OUT_BUFFER) {
-      size_t buffer_size;
-      if (JXL_DEC_SUCCESS !=
-          JxlDecoderImageOutBufferSize(dec, &format, &buffer_size)) {
-        fprintf(stderr, "JxlDecoderImageOutBufferSize failed\n");
-        break;
-      }
-      if (buffer_size != *stride * *ysize) {
-        fprintf(stderr, "Invalid out buffer size %zu %zu\n", buffer_size,
-                *stride * *ysize);
-        break;
-      }
-      size_t pixels_buffer_size = buffer_size * sizeof(uint8_t);
-      pixels = malloc(pixels_buffer_size);
-      void *pixels_buffer = (void *)pixels;
-      if (JXL_DEC_SUCCESS != JxlDecoderSetImageOutBuffer(dec, &format,
-                                                         pixels_buffer,
-                                                         pixels_buffer_size)) {
-        fprintf(stderr, "JxlDecoderSetImageOutBuffer failed\n");
-        break;
-      }
-    } else if (status == JXL_DEC_FULL_IMAGE) {
-      // This means the decoder has decoded all pixels into the buffer.
-      success = 1;
-      break;
-    } else if (status == JXL_DEC_SUCCESS) {
-      fprintf(stderr, "Decoding finished before receiving pixel data\n");
-      break;
-    } else {
-      fprintf(stderr, "Unexpected decoder status: %d\n", status);
-      break;
-    }
-  }
-  JxlDecoderDestroy(dec);
-  if (success) {
-    return pixels;
-  } else {
-    free(pixels);
-    return NULL;
-  }
-}
-
-static GdkPixbuf *gdk_pixbuf__jxl_image_load(FILE *f, GError **error) {
-  size_t data_size;
-  int status;
-  gpointer data;
+static gpointer begin_load(GdkPixbufModuleSizeFunc size_func,
+                           GdkPixbufModulePreparedFunc prepare_func,
+                           GdkPixbufModuleUpdatedFunc update_func,
+                           gpointer user_data, GError **error) {
+  GdkPixbufJXLDecoderState *decoder_state =
+      g_malloc(sizeof(GdkPixbufJXLDecoderState));
+  memset(decoder_state, 0, sizeof(*decoder_state));
```

Comment:
Is it necessary to zero this memory first if you immediately populate it?

Reply:
Well, it's not a lot of memory - but it's there for (a) initializing some struct members that are not initialized otherwise (no member initializers in C...) and (b) because in case of early failures, the code in the cleanup goto will free up NULLs rather than uninitialized memory (which would be bad)

Reply:
If it is, then `g_malloc0()` does exactly this. Calls `calloc()` internally.

Reply:
Ah, very useful to know, thanks! Done.